[SPARK-49566][SQL] Add SQL pipe syntax for the SET operator

dtenedor opened the pull request:
### What changes were proposed in this pull request?

This PR adds SQL pipe syntax support for SET operator.

This operator removes one or more existing column from the input table and replaces each one with a new computed column whose value is equal to evaluating the specified expression.

This is equivalent to `SELECT * EXCEPT (name), <newExpressions> AS name` in the SQL compiler. It is provided as a convenience feature and some functionality overlap exists with lateral column aliases.

For example:

```
-- Setting with an expression.
values (0, 'pqr', 2), (3, 'tuv', 5) as tab(a, b, c)
|> set c = a + length(b);

0, 'pqr', 3
3, 'tuv', 6
```

### Why are the changes needed?

The SQL pipe operator syntax will let users compose queries in a more flexible fashion.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds a few unit test cases, but mostly relies on golden file test coverage. I did this to make sure the answers are correct as this feature is implemented and also so we can look at the analyzer output plans to ensure they look right as well.

### Was this patch authored or co-authored using generative AI tooling?

No

dtenedor committed:
commit

dtenedor commented:
cc @cloud-fan @gengliangwang here is the `|> SET` operator :)

cloud-fan commented on sql/core/src/test/resources/sql-tests/results/pipe-operators.sql.out:
```diff
@@ -678,6 +678,156 @@ org.apache.spark.sql.AnalysisException
+table t
+|> set x = 1
+-- !query schema
+struct<y:string,x:int>
```
so the new columns in SET are always put at the end instead of in-place update?

dtenedor replied:
Good catch. I was using "SELECT * EXCEPT" and appending the new expression at the end to implement the behavior. It turns out this changed the column order, so I updated the `UnresolvedStarExcept` expression to support performing the replacement in the original column position instead. This expression is now named `UnresolvedStarExceptOrReplace`.

Note that outside of SQL pipe operators, we could choose to use this to support "SELECT * EXCEPT REPLACE" syntax [1] in the future if we want. I am not implementing that now in the PR in the interest of keeping the PRs small and focused, but leaving a note here that we could do it in the future if desired.

[1] https://stackoverflow.com/questions/73514343/databases-that-support-select-except-replace

dtenedor committed:
respond to code review comments

respond to code review comments

respond to code review comments

respond to code review comments

dtenedor reviewed (commented):
Thanks @cloud-fan for your review! Please take another look.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:
```diff
@@ -462,13 +462,15 @@ abstract class Star extends LeafExpression with NamedExpression {
- * and SELECT * EXCEPT ([[UnresolvedStarExcept]]). [[UnresolvedStar]] is just a case class of this,
- * while [[UnresolvedStarExcept]] adds some additional logic to the expand method.
-  */
+ *               This class provides the shared behavior between the classes for
```
This is the class doc instead of the parameter doc of `target`, why do we change the indentation here?

dtenedor replied:
Good catch, I think IntelliJ accidentally auto-reformatted this. I fixed it back to the original indentation.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -5978,7 +5982,56 @@ class AstBuilder extends DataTypeAstBuilder
+        val projectList: Seq[NamedExpression] =
+          Seq(UnresolvedStarExceptOrReplace(
+            target = None, excepts = Seq(Seq(ident)), replacements = Some(Seq(replacement))))
+        plan = Project(projectList, plan)
```
Why do we create a stack of `Project` here?  Do we want to support "LCA" in SET? e.g. `SET a=b, b=c`.

dtenedor committed:
respond to code review comments

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Thanks again @cloud-fan for your reviews!!

gengliangwang commented on sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:
```diff
@@ -1520,6 +1521,15 @@ operatorPipeRightSide
     ;
 
+operatorPipeSetAssignmentSeq
+    : ident+=errorCapturingIdentifier
```
`errorCapturingIdentifier` can be quoted identifiers. Shall we also test 
```
set `a.b` = ...
```
?

dtenedor replied:
Good idea, added a test case for this.

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -5978,7 +5982,59 @@ class AstBuilder extends DataTypeAstBuilder
+          "SQL pipe syntax |> SET operator with an alias assigned with [AS] aliasName", ctx)
+      case (ident, target) =>
+        // Check uniqueness of the assignment keys.
+        val checkKey = if (SQLConf.get.caseSensitiveAnalysis) {
```
Do we have a test case for this one?

dtenedor replied:
Sounds good, I added unit tests covering this feature with the config enabled and disabled.

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Thanks @gengliangwang for your review!

cloud-fan commented on sql/core/src/test/resources/sql-tests/inputs/pipe-operators.sql:
```diff
@@ -241,6 +241,85 @@ table t
+|> extend 3 as c
+|> set a = b, b = c;
+
+-- Setting two times with a lateral reference.
```
how about self reference? e.g. `SET c = c + 1`, shall we allow it?

cloud-fan replied:
further more, if we allow it, do we need the SET keys uniqueness check? People may want to do `SET c = c + 1, c = c * 2`, and this can be well supported because we use one `Project` per SET key.

dtenedor replied:
You're right, both of these are possible now. I removed the duplicate key assignment check and these cases work now. I added a couple test cases to cover it.

dtenedor committed:
respond to code review comments

respond to code review comments

dtenedor reviewed (commented):
Thanks @cloud-fan for your review again.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:
```diff
@@ -571,8 +571,15 @@ abstract class UnresolvedStarBase(target: Option[Seq[String]]) extends Star with
-case class UnresolvedStarExcept(target: Option[Seq[String]], excepts: Seq[Seq[String]])
+case class UnresolvedStarExceptOrReplace(
+    target: Option[Seq[String]],
+    excepts: Seq[Seq[String]],
```
The `excepts` here can be nested columns, so technically we can support nested columns in the SET clause. This can be a followup work as an improvement.

dtenedor replied:
@cloud-fan yes, and also we could support `SELECT * REPLACE(<column_name>, <new_expression>)` as well (some other SQL engines support this). We can keep these features in mind for adding to Spark in the future.

cloud-fan commented:
thanks, merging to master!

cloud-fan commented:
@HyukjinKwon @allisonwang-db  shall we implement `df.withColumn` with this way?
